### ISSUE TITLE
More tighter integration of OpenAI models with ML.NET

### ISSUE THREAD

**Issue by VenkateshSrini** [dotnet/machinelearning#6600]
**Is your feature request related to a problem? Please describe.**
MIcrosoft has a strong alliance with OpenAI. Our expectation is that we should be able to easily get these models say from Model Zoo of Azure (similar HuggingFace) and use them in our application.

**Describe the solution you'd like**
1) Download and Save a GPT model in a very compressed format
2) on the first access, uncompress and cache the model file. I expect it to be huge
3) Like the GPT API provided by open AI, provide precise input and output objects.
4) Source application will fill in the request object and invoke the model as if it was an resource like database. It will receive the response which it will process

**Describe alternatives you've considered**
1) Calling API of OpenAI
2) Trying to use model from HuggingFace. I did not how to do this

**Comment by JeremyMahieu** [comment]
Do you really think OpenAI will release their models just so it could be used in ML.NET? Also, it's probably not feasable to run the biggest models on normal hardware, you probably need >700GB of RAM.

I don't think a request for OpenAI to release their model belongs here.

**Comment by JeremyMahieu** [comment]
Do you really think OpenAI will release their models just so it could be used in ML.NET? Also, it's probably not feasable to run the biggest models on normal hardware, you probably need >700GB of RAM.

I don't think a request for OpenAI to release their model belongs here.

**Comment by VenkateshSrini** [comment]
@JeremyMahieu ,
I understand that one of the successful aspects of OpenAI is, it's capability to run a model in remote machines and get the result to us. So Since OpenAI is totally different set of options on ML plane, My ask is, Can we have something like ML.NET.OpenAI which specializes and works closely with OpenAI. It should provide day 1 support for new release of open AI models like GPT-4 or say Whisper OpenAI or ChatGPT AI rather than adding these capabilities to Azure SDK

**Comment by JeremyMahieu** [comment]
So an API Wrapper as nuget package? Check out https://github.com/OkGoDoIt/OpenAI-API-dotnet

**Comment by VenkateshSrini** [comment]
Yes I'm aware of this SDK. But what about the frameworks like Whisper OpenAI they need additional support which I think can be accommodate if we bring it in the umbrella of ML.NET. The things will grow much complex in future when you could do OCR kind of activity with OpenAI.

**Comment by luisquintanilla** [comment]
Hi all,

Thanks for the feedback and suggestions. A few questions:

- What scenarios are you looking to leverage OpenAI for?
- What are some gaps that you think tighter integration between OpenAI and ML.NET provide?

**Comment by VenkateshSrini** [comment]
> Hi all,
> 
> Thanks for the feedback and suggestions. A few questions:
> 
> * What scenarios are you looking to leverage OpenAI for?
> * What are some gaps that you think tighter integration between OpenAI and ML.NET provide?

**What scenarios are you looking to leverage OpenAI**

1. Answer questions based on an uploaded document (something like Lab test report). It should also use already existing knowledge in model in simple normal English
2. Summarize the contents in bunch of knowledge uploaded. It may be a serious of prescriptions, test in a simple English
3. Convert speech to text. Basically transcribe the instructions of Practioner etc:- 

**What are some gaps that you think tighter integration between OpenAI and ML.NET provide?**
1) ML.NET needs to be an umbrella product. What ever it does today is fine and it can enhance as it goes. It should also have the capability to invoke open API chat gpt, gpt - 3.5 or gpt - 4. These are just REST API wrapper. For GPT-4 invocation we still do not have a proper SDK. Apart from that we would also need to invoke WHISPER open AI. Today even for pithing we need to install addition s./w to it. We need all these nuisances to be covered in ML.NET. Dev's needs to just include one single library and just start working with the audio inputs.

**Comment by geffzhang** [comment]
ml.net for tighter integration with openAI, more support can be provided through the [Semantic Kernel](https://github.com/microsoft/semantic-kernel)

**Comment by Darlanio** [comment]
I would like to see the possibility to train NLP/GPT/LLM in ML.NET and then use it, yes.
To do this on a local machine would take ages if the corpus and the model are too big, but it is possible to train in Azure and use the model (influence?) locally with acceptable speed (a few tokens per second on smaller models).

ML.NET should be able to
- Accept a file as the corpus
- Create the model
- Train the model (which will probably take ages on a local machine, but anyways)
- Optionally train the model using the corpus in Azure platform (will cost _some_ money, but will be quicker).
- Use the model locally with ML.NET to produce predictions on "what will come next" after a certain text/prompt.

Should not be difficult to implement for the savvy, but would be slow if the model/corpus is too large.

**Comment by geffzhang** [comment]
[Semantic Kernel](https://github.com/microsoft/semantic-kernel) can't native integration with Hugging Face, ML.NET has no native C# support for Hugging Face local model.
